### DISCUSSION TITLE
Why auth request can be accessed only by guests?

### DISCUSSION THREAD

**Discussion by teenjuna** [pocketbase/pocketbase#1494] [Q&A] [answered]
Hi! I'm trying to create an `OnRecordAuthRequest` hook that checks if the client is already authorized as another user and does some data merging in that case. However, it seems like PocketBase prevents authorized clients from reaching authorization endpoints. Is there a good reason for this? How can I disable this behavior?

**Comment by ganigeorgiev** [comment, marked as answer]
The `/apis/collections/:collection/auth-with-password` endpoint requires the user to be "guest", aka. to NOT have `Authorization: TOKEN` header with the request.

> How can I disable this behavior?

There is no easy way to change it. One workaround could be to [register a middleware](https://pocketbase.io/docs/router/#middlewares) and intercept the `/auth-with-password` action.

>  Is there a good reason for this?

It just seems logical to me. Additionally it also prevents accidentally "overwriting" your `pb.authStore` state after a successful auth call.

I can change it, but I'm not sure that I understand the use case. Could you elaborate a little more what you are trying to do and what is the end goal (pseudo-code is also ok)?

**Comment by teenjuna** [comment]
My use case is the transition between "anonymous" user and an authenticated user in e-commerce app. When anonymous user logs in, I want to transfer its shopping cart & orders data to the new account. In order to do that, I need to extract the information about the current user (if it exists at all):
```go
app.OnRecordAuthRequest().Add(func(e *core.RecordAuthEvent) error {
	// Extract current user.
	user, _ := e.HttpContext.Get(apis.ContextAuthRecordKey).(*models.Record)
	if user == nil {
		return nil
	}

	if user.Id != e.Record.Id {
		log.Printf("MOVING FROM USER %s TO USER %s\n", user.Id, e.Record.Id)
	}

	return nil
})

```

**Reply by ganigeorgiev** [reply]
Thanks for providing more context. I'll consider removing the `RequireGuestOnly()` middleware from the endpoint in the upcoming v0.11 or together with https://github.com/pocketbase/pocketbase/issues/1240.

**Reply by teenjuna** [reply]
Thank you!

**Reply by teenjuna** [reply]
@ganigeorgiev by the way, do you have an approximate timeline for the v0.11 release?

**Reply by ganigeorgiev** [reply]
I'll try to publish a v0.11.0 (or maybe a prelease because I'm still not sure about some of the changes) by the end of this week, but there are no fixed ETAs.

**Reply by teenjuna** [reply]
Hey @ganigeorgiev! Sorry to bother you again. I just tried to do the thing I described in this thread, but it seems like `js-sdk` doesn't send `Authorization` header when `.authWithPassword` is called, therefore the `user` in the code above is `nil`...

**Reply by ganigeorgiev** [reply]
The support for sending the existing auth token with the `authWithPassword` request was added in [v0.10.0 of the SDK](https://github.com/pocketbase/js-sdk/releases/tag/v0.10.0).
If you are already using the latest SDK version, please provide information on your setup and a code sample how you are sending the requests and I'll investigate it in more details.

**Reply by teenjuna** [reply]
You're right, I completely forgot to update `js-sdk` along with the `PocketBase` itself. My fault, it works. Sorry!

**Reply by teenjuna** [reply]
@ganigeorgiev, on further tackling, it seems like `authWithPassword` sends `Authorization` for the new user, not the previous one.

Here is what I'm doing in my login action on frontend:
```typescript
await pb
  .collection(Collections.Users)
  .authWithPassword(values.email, values.password);

// For SSR
const cookie = pb()!.authStore.exportToCookie({ httpOnly: false });
cookies.set(AUTH_COOKIE_KEY, cookie, { path: "/", expires: 30 });

window.location.href = "/";
```

And here is how I'm trying to handle it on backend:
```go
app.OnRecordAuthRequest().Add(func(e *core.RecordAuthEvent) error {
        // Extract current user.
	user, _ := e.HttpContext.Get(apis.ContextAuthRecordKey).(*models.Record)
	// For debug purposes
	if user != nil {
		log.Printf("Previous user is %s, current user id %s", user.Id, e.Record.Id)
	}
	if user == nil || user.Id == e.Record.Id || user.Email() != "" {
		return nil
	}
	
        // ...
 })
```

This debug print shows me that `e.Record.Id` is always the same as `user.Id`

**Reply by ganigeorgiev** [reply]
I'm sorry, I don't understand what `pb()` is in the above. If you are sending the following 2 requests:

```js
await pb.collection('users').authWithPassword('old@example.com', '123456')

// the second auth request will have Authorization: TOKEN with it
await pb.collection('users').authWithPassword('new@example.com', '123456')
```

Then in the `OnRecordAuthRequest()` hook `e.HttpContext.Get(apis.ContextAuthRecordKey).(*models.Record)` should refer to the first user.

**Reply by teenjuna** [reply]
I created a small client-only demo and it works. Looks like I messed up with the SSR. Sorry again, I will try to do more research before writing here in the future.
